File: android/src/main/java/com/synerise/sdk/react/inapp/inline/RNInlineInAppViewRegistry.java
package com.synerise.sdk.react.inapp.inline;

import androidx.annotation.Nullable;

import com.synerise.sdk.injector.inapp.ui.inline.InlineInAppView;

import java.util.Iterator;
import java.util.Map;
import java.util.Objects;
import java.util.HashMap;
import java.util.concurrent.atomic.AtomicLong;

public final class RNInlineInAppViewRegistry {

    private static final Map<String, RetainedInlineView> retainedViews = new HashMap<>();
    private static final AtomicLong idGenerator = new AtomicLong();

    private RNInlineInAppViewRegistry() {
    }

    public static String retain(String placementKey, InlineInAppView view) {
        synchronized (retainedViews) {
            releaseRetainedViewsForPlacementKey(placementKey);
            String viewId = "inline-" + idGenerator.incrementAndGet();
            retainedViews.put(viewId, new RetainedInlineView(placementKey, view));
            return viewId;
        }
    }

    @Nullable
    static InlineInAppView takeRetainedView(String viewId) {
        synchronized (retainedViews) {
            RetainedInlineView retained = retainedViews.remove(viewId);
            return retained != null ? retained.view : null;
        }
    }

    /**
     * A newer campaign for the same placement supersedes anything still waiting to be mounted,
     * so the older view is released rather than left holding a WebView forever.
     */
    private static void releaseRetainedViewsForPlacementKey(String placementKey) {
        Iterator<Map.Entry<String, RetainedInlineView>> entries = retainedViews.entrySet().iterator();
        while (entries.hasNext()) {
            RetainedInlineView retained = entries.next().getValue();
            if (Objects.equals(retained.placementKey, placementKey)) {
                entries.remove();
                retained.view.release();
            }
        }
    }

    private static final class RetainedInlineView {
        final String placementKey;
        final InlineInAppView view;

        RetainedInlineView(String placementKey, InlineInAppView view) {
            this.placementKey = placementKey;
            this.view = view;
        }
    }
}

File: ios/ReactNativeSynerise/Modules/RNInlineInAppViewRegistry.m
//
//  RNInlineInAppViewRegistry.m
//  react-native-synerise-sdk
//
//  Created by Synerise
//  Copyright © 2021 Synerise. All rights reserved.
//

#import "RNInlineInAppViewRegistry.h"

#import <SyneriseSDK/SyneriseSDK.h>

NS_ASSUME_NONNULL_BEGIN

@implementation RNInlineInAppViewRegistry

static NSMutableDictionary<NSString *, SNRInlineInAppView *> *retainedViews;
static NSMutableDictionary<NSString *, NSString *> *placementKeysByIdentifier;
static NSUInteger identifierCounter;

+ (void)initialize {
    if (self == [RNInlineInAppViewRegistry class]) {
        retainedViews = [NSMutableDictionary dictionary];
        placementKeysByIdentifier = [NSMutableDictionary dictionary];
        identifierCounter = 0;
    }
}

+ (NSString *)retainInlineInAppView:(SNRInlineInAppView *)view placementKey:(nullable NSString *)placementKey {
    @synchronized (self) {
        [self releaseRetainedViewsForPlacementKey:placementKey];

        identifierCounter += 1;
        NSString *viewIdentifier = [NSString stringWithFormat:@"inline-%lu", (unsigned long)identifierCounter];
        retainedViews[viewIdentifier] = view;
        if (placementKey != nil) {
            placementKeysByIdentifier[viewIdentifier] = placementKey;
        }

        return viewIdentifier;
    }
}

+ (nullable SNRInlineInAppView *)takeRetainedViewWithIdentifier:(NSString *)viewIdentifier {
    @synchronized (self) {
        SNRInlineInAppView *view = retainedViews[viewIdentifier];
        [retainedViews removeObjectForKey:viewIdentifier];
        [placementKeysByIdentifier removeObjectForKey:viewIdentifier];

        return view;
    }
}

/**
 * A newer campaign for the same placement supersedes anything still waiting to be mounted, so the
 * older view is dropped rather than left retained forever.
 */
+ (void)releaseRetainedViewsForPlacementKey:(nullable NSString *)placementKey {
    if (placementKey == nil) {
        return;
    }

    NSArray<NSString *> *identifiers = [placementKeysByIdentifier allKeys];
    for (NSString *identifier in identifiers) {
        if ([placementKeysByIdentifier[identifier] isEqualToString:placementKey] == YES) {
            SNRInlineInAppView *staleView = retainedViews[identifier];
            staleView.delegate = nil;
            [staleView removeFromSuperview];
            [retainedViews removeObjectForKey:identifier];
            [placementKeysByIdentifier removeObjectForKey:identifier];
        }
    }
}

@end

NS_ASSUME_NONNULL_END
